paddle/framework/operator.h

```diff
@@ -317,26 +318,108 @@ class ExecutionContext : public InferShapeContext {
   const platform::DeviceContext& device_context_;
 };
 
+class CompileTimeInferShapeContext : public InferShapeContextBase {
+ public:
+  CompileTimeInferShapeContext(const OpDescBind& op, const BlockDescBind& block)
+      : op_(op), block_(block) {}
+
+  bool HasInput(const std::string& name) const override {
+    const std::vector<std::string>& input_names = op_.Input(name);
+    PADDLE_ENFORCE_EQ(input_names.size(), 1UL, "Inputs(%s) length is not 1",
+                      name);
+    return block_.HasVar(input_names[0]);
+  }
+
+  bool HasOutput(const std::string& name) const override {
+    const std::vector<std::string>& output_names = op_.Output(name);
+    PADDLE_ENFORCE_EQ(output_names.size(), 1UL, "Outputs(%s) length is not 1",
+                      name);
+    return block_.HasVar(output_names[0]);
+  }
+
+  bool HasInputs(const std::string& name) const override {
```

Comment:
Most of the Time, One input will only have one value, 
```python
X = [x1]
```
`HasInput` has a check that for X, there should be and only be one value.

for `HasInputs`, it just require the length is not zero because the length is not certain.

Reply:
I don't understand what does the example `X=[x1]` mean?

Reply:
Also, what's the difference between `HasInput` and `HasInputs`?  From the English language, it seems that they would have the same functionality.

Reply:
In our current implementation, all inputs outputs are repeated filed in [protobuf](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/framework.proto#L54),  so the real data sturcture for input/output is:
```
VarName = [v1, v2, v3 ...]
```
Most of the time, One VarName only have one corresponding variable, but in some case, the input length is not certain, for example:    `mul_op` will have data like:
```python
X = [x1, x2, x3, ...]
Y = [y1, y2, y3, ...]
Out = [x1+y1, x2+y2, x3+y3, ...]
```